src/Nethermind/Nethermind.Evm/Precompiles/Bls/SubgroupChecks.cs

```diff
@@ -0,0 +1,135 @@
+    internal static readonly Fp2 S = new(new BigInteger([0x13, 0x52, 0x03, 0xE6, 0x01, 0x80, 0xA6, 0x8E, 0xE2, 0xE9, 0xC4, 0x48, 0xD7, 0x7A, 0x2C, 0xD9, 0x1C, 0x3D, 0xED, 0xD9, 0x30, 0xB1, 0xCF, 0x60, 0xEF, 0x39, 0x64, 0x89, 0xF6, 0x1E, 0xB4, 0x5E, 0x30, 0x44, 0x66, 0xCF, 0x3E, 0x67, 0xFA, 0x0A, 0xF1, 0xEE, 0x7B, 0x04, 0x12, 0x1B, 0xDE, 0xA2], true, true), new BigInteger([0x06, 0xAF, 0x0E, 0x04, 0x37, 0xFF, 0x40, 0x0B, 0x68, 0x31, 0xE3, 0x6D, 0x6B, 0xD1, 0x7F, 0xFE, 0x48, 0x39, 0x5D, 0xAB, 0xC2, 0xD3, 0x43, 0x5E, 0x77, 0xF7, 0x6E, 0x17, 0x00, 0x92, 0x41, 0xC5, 0xEE, 0x67, 0x99, 0x2F, 0x72, 0xEC, 0x05, 0xF4, 0xC8, 0x10, 0x84, 0xFB, 0xED, 0xE3, 0xCC, 0x09], true, true));
+
+    [SkipLocalsInit]
+    public static bool G1IsInSubGroup(ReadOnlySpan<byte> p)
+    {
+        Span<byte> buf = stackalloc byte[3 * 128];
+        Span<byte> mulArgs = buf[128..(2 * 128 + 32)];
+        Span<byte> addArgs = buf[..(2 * 128)];
+        Span<byte> res = buf[(2 * 128)..];
+
+        p.CopyTo(mulArgs);
+        G1SubgroupMultiplier.CopyTo(mulArgs[128..]);
+
+        if (!Pairings.BlsG1Mul(mulArgs, addArgs[..128]))
+        {
+            return false;
+        }
+
+        Phi(p, addArgs[128..]);
+
+        if (!Pairings.BlsG1Add(addArgs, res))
+        {
+            return false;
+        }
+
+        return res.IndexOfAnyExcept((byte)0) == -1;
+    }
+
+    [SkipLocalsInit]
+    public static bool G2IsInSubGroup(ReadOnlySpan<byte> p)
+    {
+        Span<byte> buf = stackalloc byte[3 * 256];
+        Span<byte> mulArgs = buf[256..(2 * 256 + 32)];
+        Span<byte> addArgs = buf[..(2 * 256)];
+        Span<byte> res = buf[(2 * 256)..];
+
+        p.CopyTo(mulArgs);
+        G2SubgroupMultiplier.CopyTo(mulArgs[256..]);
+        if (!Pairings.BlsG2Mul(mulArgs, addArgs[..256]))
+        {
+            return false;
+        }
+
+        Psi(p, addArgs[256..]);
+
+        if (!Pairings.BlsG2Add(addArgs, res))
+        {
+            return false;
+        }
+
+        return res.IndexOfAnyExcept((byte)0) == -1;
+    }
+
+    internal static void Phi(ReadOnlySpan<byte> p, Span<byte> res)
+    {
+        BigInteger x = new(p[..64], true, true);
+        x = Beta * x % BlsBaseFieldOrder;
+        x.ToBigEndianByteArray(64).CopyTo(res);
+        p[64..].CopyTo(res[64..]);
+    }
+
+    internal static void Psi(ReadOnlySpan<byte> p, Span<byte> res)
+    {
+        Fp2 x = new Fp2(p[..128]).Conjugate() * R;
+        Fp2 y = new Fp2(p[128..]).Conjugate() * S;
+        x.ToBytes(res[..128]);
+        y.ToBytes(res[128..]);
+    }
+
+    internal static BigInteger Normalise(in BigInteger x)
+    {
+        BigInteger unnormalised = x % BlsBaseFieldOrder;
```

Comment:
BigIntegers are painfully slow, do we need simple UInt512 types?

Reply:
At some point the subgroup checks should be implemented in the cryptography library, so probably not worth optimising for now